### PR TITLE
Add robot mode publisher, masterboard publisher and waitForURProgram example function

### PR DESCRIPTION
**Pull request by felixvd**
When sending a script to the robot, it is often necessary to know when it has stopped. @gavanderhoorn mentioned [here](https://github.com/ThomasTimm/ur_modern_driver/issues/162) that "there's no way to detect whether 'the script' has finished", but this is implemented in other popular python libraries like [urx](https://github.com/SintefRaufossManufacturing/python-urx) via the program_running status variable.

The ur_modern_driver already parses the parameter, but does not expose it. This PR publishes the parameter to a topic `/ur_driver/program_running`, and adds a simple example that allows C++ users to wait for the program on the robot to finish.

If the wait function is called when the regular ur_modern_driver is running on the controller (i.e. when a goal is sent through the action interface), it will not terminate. Some users might get confused about this, but the functionality is required for some use cases (including ours), and I believe the message that appears on-screen is informative enough.

[A related ROS Answers thread](https://answers.ros.org/question/285345/action-completion-from-ur-controller/).

### CODE REVIEW

**Comment by gavanderhoorn** [comment]
@felixvd wrote:
> When sending a script to the robot, it is often necessary to know when it has stopped. @gavanderhoorn mentioned [here](https://github.com/ThomasTimm/ur_modern_driver/issues/162) that "there's no way to detect whether 'the script' has finished", but this is implemented in other popular python libraries like [urx](https://github.com/SintefRaufossManufacturing/python-urx) via the program_running status variable.

my comment was specifically made in the context of programs running in parallel to the driver or as a replacement. As the driver is a program as well, there is no clear way to know which program is running or not. You mention this ambiguity yourself already:

> If the wait function is called when the regular ur_modern_driver is running on the controller (i.e. when a goal is sent through the action interface), it will not terminate. 

I'm not so sure I agree with your next statement:

> Some users might get confused about this, but the functionality is required for some use cases (including ours), and I believe the message that appears on-screen is informative enough.

`ROS_INFO` is not always shown on the console, and the nr of users/devs that don't check the various ROS logging tools is too high to rely on this alone.

It would perhaps be better to provide a `isProgramRunning()` function, which could then be used by clients in a simple while-loop (like in `wait_for_program.h`). This would avoid introducing a potentially infinite wait, or at least in driver provided code.

**Comment by gavanderhoorn** [comment]
Just a clarification: I'm not at all against exposing this information to the ROS side of things.

I am just wondering whether it would not make more sense to just expose the entirety of the `MasterBoard` and `RobotMode` and friends: interpretation of those msgs is completely tied to UR robots anyway, so code dealing with those messages will be tied to those platforms as well.

I'm not a fan at all of exposing such implementation details from a robot driver, but right now we don't have any abstractions in place that we could use to avoid it (for many robots: "is a program running?" is a strange thing to ask anyway, as there are always programs running).

https://github.com/ThomasTimm/ur_modern_driver/pull/181 introduces publishers for all of that data, so what about porting that PR to this fork and then providing an implementation of `waitForURProgram(..)` that just checks the RobotMode field, instead of adding yet another publisher for a single boolean?

Would that work?

**Comment by felixvd** [comment]
I agree with what you said about the abstraction - I don't see a good way to represent this either. It's very specific to this robot. I didn't see @akihikoy 's PR before this, but I think it does make sense to expose all of the status data. 

I put in most of the basic functionality, but as it says in the Readme, only the fields that are shared between the different robot firmwares are published. I did not want to deal with the case distinctions right now.

I liked your `isProgramRunning()` suggestion more than just the wait function, so I put both in.

On first glance, this worked fine on our robot, but I didn't run any tests.

**Comment by felixvd** [comment]
I tested the waitForURProgram() function on our UR5 today without issue.

I noticed that the behavior of `isProgramRunning()` is unintuitive when mistyping the namespace, though, so I changed it to return true if no robot message is received. This way, a namespace mistake only results in more waiting, rather than the program continuing. This should be both safer and more understandable for the user, since the error message also gets printed to the screen every 2 seconds.

**Comment by gavanderhoorn** [comment]
> I noticed that the behavior of `isProgramRunning()` is unintuitive when mistyping the namespace, though, so I changed it to return true if no robot message is received. This way, a namespace mistake only results in more waiting, rather than the program continuing. 

Why not throw an exception?

That is a relatively common pattern in ROS and a lot more explicit about what is going on.

**Comment by felixvd** [comment]
I'm just never sure about good practices with exceptions so I tend not to touch them. I updated it to throw one.

File: include/ur_modern_driver/wait_for_program.h [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+// Note: topic_namespace needs a leading slash.
+bool waitForURProgram(std::string topic_namespace = "", ros::Duration timeout = ros::Duration(60.0))
+{
+  ROS_INFO("Waiting for UR program to finish. Only run this after sending custom URScripts and not the regular motion commands, or this call will not terminate before the timeout.");
+  ros::Time t_now, t_start = ros::Time::now();
+  while (isProgramRunning(topic_namespace))
+  {
+    ros::Duration(.05).sleep();
+  }
```
Comment (gavanderhoorn) [inline include/ur_modern_driver/wait_for_program.h:32]: Is the `timeout` actually used here?

File: include/ur_modern_driver/wait_for_program.h [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+#include "ros/ros.h"
+#include "ur_modern_driver/RobotModeDataMsg.h"
+// Do not forget to add ur_modern_driver to your CMakeLists.txt
+
+// Note: topic_namespace needs a leading slash.
+bool isProgramRunning(std::string topic_namespace = "")
+{
+  boost::shared_ptr<ur_modern_driver::RobotModeDataMsg const> sharedPtr;
+  ur_modern_driver::RobotModeDataMsg robot_mode_state;
```
Comment (gavanderhoorn) [inline include/ur_modern_driver/wait_for_program.h:10]: We're using C++11, so perhaps we should do something like (haven't tested it):

```C++
auto msg = ros::topic::waitForMessage<ur_modern_driver::RobotModeDataMsg>(..);
if (msg)
  return msg->is_program_running;
else
{
  ROS_ERROR(..);
  throw ros::Exception(..);
}
```

**Comment by gavanderhoorn** [comment]
My apologies (and gratitude) for iterating this many times. :+1:

File: include/ur_modern_driver/wait_for_program.h (hunk 1 above)
Comment (felixvd) [reply include/ur_modern_driver/wait_for_program.h:32]: Whoops.

File: include/ur_modern_driver/wait_for_program.h (hunk 2 above)
Comment (felixvd) [reply include/ur_modern_driver/wait_for_program.h:10]: Sure. It works, so I added it in like that (to the next commit).

**Comment by felixvd** [comment]
Cheers, thanks for the review.

**Comment by gavanderhoorn** [comment]
I'll port this to `ros-industrial/ur_modern_driver` when I get a chance.

Or if you feel like it @felixvd you could re-open against the main repository now :)

Be sure to select `kinetic-devel` as the target branch.

**Comment by felixvd** [comment]
Rebased on kinetic-devel and opened against the main repository, so I'll close this.

**Pull request closed without merging**